### PR TITLE
*: add check metrics in it test

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve? 

- None

### What is changed and how it works?

- add check metrics  in some common test case

### Check List 

Tests 

 - Integration test

Code changes

- [x] full_mode
- [x] all_mode
- [x] only_dml
- [x] sharding
- [x] online_ddl
- [x] new_relay
- [x] dm_syncer
- [x] gtid
- [x] ha

Side effects

- None

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: tests/all_mode/run.sh [hunk 1]
```diff
@@ -83,6 +85,7 @@ function test_query_timeout() {
 	cp $cur/conf/dm-task.yaml $WORK_DIR/dm-task.yaml
 	sed -i "s/name: test/name: $ILLEGAL_CHAR_NAME/g" $WORK_DIR/dm-task.yaml
 	dmctl_start_task "$WORK_DIR/dm-task.yaml" "--remove-meta"
+	check_metric $WORKER1_PORT "dm_worker_task_state{source_id=\"mysql-replica-01\",task=\"$ILLEGAL_CHAR_NAME\"}" 3 1 31
```
Comment (lance6716) [inline tests/all_mode/run.sh:88]:
```suggestion
	check_metric $WORKER1_PORT "dm_worker_task_state{source_id=\"mysql-replica-01\",task=\"$ILLEGAL_CHAR_NAME\"}" 3 1 3
```

File: tests/full_mode/run.sh [hunk 2]
```diff
@@ -165,6 +165,10 @@ function run() {
 
 	run_dm_master $WORK_DIR/master $MASTER_PORT $cur/conf/dm-master.toml
 	check_rpc_alive $cur/../bin/check_master_online 127.0.0.1:$MASTER_PORT
+	# check dm-master metrics
+	check_metric $MASTER_PORT 'start_leader_counter' 3 0 2
+	echo "check master metric: [start_leader_counter] done"
```
Comment (lance6716) [inline tests/full_mode/run.sh:170]: this line is for debug?

File: tests/only_dml/run.sh [hunk 3]
```diff
@@ -49,31 +49,46 @@ function run() {
...
 
+	# check dm-workers metrics unit: relay file index must be 1.
+	check_metric $WORKER1_PORT "dm_relay_binlog_file" 3 0 2
+	check_metric $WORKER2_PORT "dm_relay_binlog_file" 3 0 2
+
 	# start a task in all mode, and when enter incremental mode, we only execute DML
 	dmctl_start_task $cur/conf/dm-task.yaml
+
+	# check task has started state=2 running
+	check_metric $WORKER1_PORT "dm_worker_task_state{source_id=\"mysql-replica-01\",task=\"$TASK_NAME\"}" 3 1 31
```
Comment (lance6716) [inline tests/only_dml/run.sh:88]:
```suggestion
	check_metric $WORKER1_PORT "dm_worker_task_state{source_id=\"mysql-replica-01\",task=\"$TASK_NAME\"}" 3 1 3
```

File: tests/only_dml/run.sh [hunk 4]
```diff
@@ -49,31 +49,46 @@ function run() {
...
+	check_metric $WORKER2_PORT "dm_worker_task_state{source_id=\"mysql-replica-02\",task=\"$TASK_NAME\"}" 3 1 31
```
Comment (lance6716) [inline tests/only_dml/run.sh:89]:
```suggestion
	check_metric $WORKER2_PORT "dm_worker_task_state{source_id=\"mysql-replica-02\",task=\"$TASK_NAME\"}" 3 1 3
```

**Review by lance6716 — COMMENTED** [review]
we should check at least all alerting rules which list here

https://docs.pingcap.com/zh/tidb-data-migration/stable/handle-alerts

could be done in another PR or other guy

File: tests/full_mode/run.sh (hunk 2 above)
Comment (Ehco1996) [reply tests/full_mode/run.sh:170]: yes, i will remove it

**Comment by lance6716** [comment]
> check all alerting rules

check metrics related to alerting rules. And we could remove unreasonable altering rules

**Comment by Ehco1996** [comment]
@lance6716  @lance6716  PTAL

File: tests/_utils/check_metric_gt_zero [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+port=$1
+metric_name=$2
+retry_count=$3
+
+shift 3
+
+counter=0
+while [ $counter -lt $retry_count ]; do
+	metric=$(curl -s http://127.0.0.1:$port/metrics | grep $metric_name | wc -l)
```
Comment (lance6716) [inline tests/_utils/check_metric_gt_zero:16]: does this script check `metric_name` exists (may with value `0`), or with a value that greater than zero?

File: tests/_utils/check_metric_gt_zero [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+# parameter 2: metric name
+# parameter 3: retry count, if check failed we will wait 1s before next retry, until retry time exceeds retry count
+
+set -eu
+
+port=$1
+metric_name=$2
+retry_count=$3
+
+shift 3
```
Comment (lance6716) [inline tests/_utils/check_metric_gt_zero:12]: seems this line is no need

File: tests/_utils/check_two_metric_equal [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+# parameter 4...: valid value list
+
+set -eu
+
+port=$1
+metric1_name=$2
+metric2_name=$3
+retry_count=$4
+
+shift 3
```
Comment (lance6716) [inline tests/_utils/check_two_metric_equal:14]: ditto

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: tests/_utils/check_metric_gt_zero (hunk 6 above)
Comment (Ehco1996) [reply tests/_utils/check_metric_gt_zero]: removed

File: tests/_utils/check_metric_gt_zero (hunk 5 above)
Comment (Ehco1996) [reply tests/_utils/check_metric_gt_zero:14]: this require return metric value greater than 0

Comment (lance6716) [reply tests/_utils/check_metric_gt_zero:14]: I guess because `wc -l`, the meaning of `metric` is count not metric value

Comment (Ehco1996) [reply tests/_utils/check_metric_gt_zero]: you're right, it's been fixed

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

File: tests/_utils/check_two_metric_equal [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+while [ $counter -lt $retry_count ]; do
+	metric1=$(curl -s http://127.0.0.1:$port/metrics | grep $metric1_name | grep -v "#" | awk '{print $2}')
+	metric2=$(curl -s http://127.0.0.1:$port/metrics | grep $metric2_name | grep -v "#" | awk '{print $2}')
+	if [ $metric1 -eq $metric2 ]; then
+		echo "matched"
+		exit 0
+	fi
+	((counter += 1))
+	echo "wait for valid metric for $counter-th time"
+	sleep 3
```
[the remark this answers is not in the document]
Comment (Ehco1996) [reply tests/_utils/check_two_metric_equal]: changed to `sleep 1`, that becaues i was debuging... <.<

File: tests/_utils/check_metric_gt_zero (hunk 5 above)
Comment (Ehco1996) [reply tests/_utils/check_metric_gt_zero]: it's ok to use `INF` but need to think about is number still great than `bash INF(2**31  -1)`

File: tests/new_relay/run.sh [hunk 9]
```diff
@@ -39,11 +40,26 @@ function run() {
...
 	check_sync_diff $WORK_DIR $cur/conf/diff_config.toml
 
 	run_sql_file $cur/data/db1.increment.sql $MYSQL_HOST1 $MYSQL_PORT1 $MYSQL_PASSWORD1
 
+	# relay task tranfer to worker1 with no error.
+	check_metric $WORKER1_PORT "dm_relay_data_corruption" 3 -1 1
+	check_metric $WORKER1_PORT "dm_relay_read_error_count" 3 -1 1
+	check_metric $WORKER1_PORT "dm_relay_write_error_count" 3 -1 1
+	# check worker relay space great than 0 9223372036854775807 is 2**63 -1
+	sleep 5
```
Comment (lance6716) [inline tests/new_relay/run.sh:60]: is this needed?

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: tests/new_relay/run.sh (hunk 9 above)
Comment (Ehco1996) [reply tests/new_relay/run.sh:60]: need  in ci test, but i don't know why,maybe another unstable test >.<

Comment (lance6716) [reply tests/new_relay/run.sh:60]: https://github.com/pingcap/dm/blob/48dacb1aed28184a325b27a41e3e9f06ae808045/relay/metrics.go#L153-L177

could add a failpoint to reduce ticker interval

Comment (Ehco1996) [reply tests/new_relay/run.sh]: added, fialpoint is so magic !!

**Review by lichunzhu — COMMENTED** [review]
/lgtm

**Pull request merged**